src/main/java/org/candlepin/subscriptions/tally/CombiningRollupSnapshotStrategy.java

```diff
@@ -88,14 +88,21 @@ public void produceSnapshotsFromCalculations(String accountNumber, OffsetDateTim
         OffsetDateTime endDateTime, Map<OffsetDateTime, AccountUsageCalculation> accountCalcs,
         DoubleBinaryOperator reductionFunction) {
 
+        if (accountCalcs.isEmpty()) {
+            // nothing to do here, return early to avoid a null swatch product ID lookup
+            return;
+        }
+
         String swatchProductId = getSwatchProductId(accountCalcs);
```

Comment:
Why do we need this if we are getting them all below?

Reply:
I feel like we should be telling produceSnapshotsFromCalculations what the finest granularity is instead of trying to determine what it is from the productId. Especially since we are chosing 1 based on what could be multiple productIds. It feels like it should be coming from the ProductProfile. The same one that was used to gather the metrics.

Reply:
Does it work now, sure! But it will surely show its teeth later on.

Reply:
Good idea!

I refactored to pass in the `finestGranularity`, which let us drop a fair number of lines :tada: